src/audio/drc/Testbench_drc_lin2dB/drc_lin2dB_fixed.c

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+
+//	drc_lin2dB_fixed.c :
+//	warning disabled to support file system
+//
+#pragma warning (disable : 4996)
+#pragma warning (disable : 4013)
+#include <stdio.h>
+#include "stdint.h"
+#include "math.h"
+#include "typdef.h"
+#include "norm.h"
+#include "testvector.h"
+#include <stdlib.h>
+#include <stddef.h>
+#include <string.h>
+int32_t drc_lin2db_fixed(int32_t linear, uint16_t idx, FILE* fd);		/* Input:Q6.26 Output:Q11.21 */
+static inline int32_t log10_fixed(int32_t x);
+
+
+
+int main()
+{
+	int32_t l_testvector, x[TEST_VECTOR], y[TEST_VECTOR]; 	/*Increase/decrease length as per need*/
+    uint16_t idx = 0;
+	int i;
+	/*int mkdir(const char* dirname);*/
+	mkdir("Results", 0777);
+	FILE* fd = fopen("Results/file.txt", "w");
+	/*fprintf(fd, " %10s| %10s|%10s|\n-----------+-----------+--------------+\n",  "idx", "testvector", "Fixlog10linear");*/
+	fprintf(fd, " %10s  %10s %10s \n", "idx", "testvector", "Fixlog10linear");
+
+    data_initialization_fixpt(x);
+    input_data_fixpt(x, y);
+	for (i = 0; i < TEST_VECTOR; i++)
+	{
+		l_testvector = y[i];
+		/* Data logging purpose_start*/
+        idx++;
+		fprintf(fd, " %10d %11li ", idx, l_testvector);
+		drc_lin2db_fixed(l_testvector, idx,fd);
+		/* Data logging purpose*/
+	}
+	fclose(fd);
+}
+
+/*
+ * function[y]= input_data_fixpt(x)
+ * Arguments    : const int32_t x[TEST_VECTOR]
+ *                int32_t y[TEST_VECTOR]
+ * Return Type  : void
+ */
+void input_data_fixpt(const int32_t x[TEST_VECTOR], int32_t y[TEST_VECTOR])
+{
+    memcpy(&y[0], &x[0], 639U * (sizeof(int32_t)));
+}
+
+/*
+ * Input is Q6.26: max 32.0
+ * Output range ~ (-inf, 30.1030); regulated to Q11.21: (-1024.0, 1024.0)
+ */
+
+inline int32_t drc_lin2db_fixed(int32_t linear, uint16_t idx, FILE* fd)		/* Data logging purpose*/
+{
+	int32_t log10_linear;
+	/* For negative or zero, just return a very small dB value. */
+	if (linear <= 0){
+		fprintf(fd, " %13.8f \n", -0.00047684);
+		return Q_CONVERT_FLOAT(-1000.0f, 21);
+	}
+	else{
+		log10_linear = log10_fixed(linear); /* Q6.26 */
+		log10_linear_log[idx] = log10_linear;		/* Dbg buffer*/
+		fprintf(fd, " %13li\n", log10_linear);
```

Comment:
It looks like you used this as the result of lin2db() function? This is not the final result, just the log10(x) result of Q6.26
The correct result should be in the next line --> q_mult(20, log10_linear, 0, 26, 21), which multiplies 20 to log10(x) and shift to Q11.21
This should be the root cause of the large error.

Reply:
Thanks ! Your observation is right.
Replace
```
fprintf(fd, " %13li\n", log10_linear);
with
fprintf(fd, " %13li\n", q_mult(20, log10_linear, 0, 26, 21));

```Then we don't see error